discovery/manager.go

```diff
@@ -153,41 +147,43 @@ func (m *Manager) cancelDiscoverers() {
 	for _, c := range m.discoverCancel {
 		c()
 	}
-	m.targets = make(map[poolKey][]*targetgroup.Group)
+	m.targets = make(map[poolKey]map[string]*targetgroup.Group)
 	m.discoverCancel = nil
 }
 
-func (m *Manager) addGroup(poolKey poolKey, tg []*targetgroup.Group) {
+func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tg != nil {
-			m.targets[poolKey] = tg
+		if tgs == nil {
+			close(done)
+			return
+		}
+
+		for _, tg := range tgs {
+			if tg != nil { // Some Discoverers send nil target group so need to check for it to avoid panics.
+				if _, ok := m.targets[poolKey]; !ok {
+					m.targets[poolKey] = make(map[string]*targetgroup.Group)
+				}
+				m.targets[poolKey][tg.Source] = tg
+			}
 		}
 		close(done)
 
 	}
 	<-done
 }
 
-func (m *Manager) allGroups(pk poolKey) map[string][]*targetgroup.Group {
+func (m *Manager) allGroups() map[string][]*targetgroup.Group {
 	tSets := make(chan map[string][]*targetgroup.Group)
 
 	m.actionCh <- func(ctx context.Context) {
-
-		// Sorting by the poolKey is needed so that we can have predictable tests.
-		var pKeys []poolKey
-		for pk := range m.targets {
-			pKeys = append(pKeys, pk)
-		}
-		sort.Sort(byProvider(pKeys))
-
 		tSetsAll := map[string][]*targetgroup.Group{}
-		for _, pk := range pKeys {
-			for _, tg := range m.targets[pk] {
-				if tg.Source != "" { // Don't add empty targets.
-					tSetsAll[pk.setName] = append(tSetsAll[pk.setName], tg)
-				}
+		for pkey, tsets := range m.targets {
+			for _, tg := range tsets {
+				// Even if the target group 'tg' is empty we still need to send it to the 'Scrape manager'
+				// to singal that is needs to stop all scrape loops for this target set.
```

Comment:
s/singal that is/signal that it/

Reply:
thanks,

I also run some tests locally with consul , k8x and static configs and can't see any problems.